File: python/problems/binary_search.py
from typing import List 
import math

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l < r:
            m = l + (r - l) // 2
            # < target when looking for first target value
            # <= target when looking for the last target value
            if nums[m] < target:
                # next point is when earliest point could be located 
                l = m + 1
            else: 
                r = m
        return l if nums and nums[l] == target else -1

File: typescript/binary_search.js
function search(nums, target) {
    let [l, r] = [0, nums.length - 1];
    while (l <= r) {
        const m = Math.floor(l + (r - l) / 2);
        if (nums[m] === target)
            return m;
        else if (nums[m] < target)
            l = m + 1;
        else if (nums[m] > target)
            r = m - 1;
    }
    return -1;
}
;
// const input = [-1,0,3,5,9,12]
// search(input, 4)
//
//
//
// 
